template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateProcessor.java

```diff
@@ -243,7 +250,39 @@ private Map<String, Set<ServiceComponent>> getNonGatewayServicesByHostGroup() {
                                 .filter(i -> !i.getComponent().equalsIgnoreCase("GATEWAY"))
                                 .collect(toSet())
                 ));
+    }
+
+    private Set<String> collectComponentByHostGroupWithRoleDependencies() {
+        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponents = getNonGatewayServicesByHostGroup();
+        return hgToNonGwServiceComponents.values().stream()
+                .filter(serviceComponents -> !Strings.isNullOrEmpty(getDependentService(serviceComponents)))
+                .map(serviceComponents -> ROLE_DEPENDENCY.get(getDependentService(serviceComponents)))
+                .collect(toSet());
+    }
+
+    private String getDependentService(Set<ServiceComponent> serviceComponents) {
+        Set<String> serviceComponentRoles = collectComponents(serviceComponents);
+        return ROLE_DEPENDENCY.keySet().stream().filter(serviceComponentRoles::contains).findFirst().orElse(null);
+    }
+
+    public List<String> getUnhealthyDependentHostGroups(StackDto stackdto) {
+            Map<String, Set<String>> nonGatewayServices = getNonGatewayComponentsByHostGroup();
+            Set<String> dependentComponent = collectComponentByHostGroupWithRoleDependencies();
+            Set<String> dependentHostGroupNames = nonGatewayServices.entrySet().stream()
+                    .filter(allComponentsInHostGroup -> isDependentComponentPresentInHostGroup(allComponentsInHostGroup.getValue(), dependentComponent))
+                    .map(Entry::getKey)
+                    .collect(toSet());
+            List<String> allUnHealthyHostGroupNames = stackdto.getAllAvailableInstances().stream()
+                    .filter(instanceMetaData -> InstanceStatus.SERVICES_UNHEALTHY.equals(instanceMetaData.getInstanceStatus()))
```

Comment:
!SERVICES_HEALTHY instead of SERVICES_UNHEALTHY. That will catch a much broader set of issues. We explicitly need SERVICES_HEALTHY to allow scaling.